plugins/ShinyForms/app/models/shiny_forms/form.rb

```diff
@@ -6,20 +6,27 @@ class Form < ApplicationRecord
     include ShinyName
     include ShinySlug
 
+    validates :slug, uniqueness: true
+
     # Instance methods
 
-    def handlers
-      @handlers ||= FormHandler.new
+    def send_to_handler( form_data )
+      return false_after_logging_warning unless handler_exists?
+
+      handlers.public_send( handler.to_sym, self, form_data )
     end
 
-    def send_to_handler( form_data )
-      if handlers.respond_to?( handler )
-        handlers.public_send( handler.to_sym, self, form_data )
-        true
-      else
-        Rails.logger.warn "Unknown form handler '#{handler}' (form ID: #{id})"
-        false
-      end
+    def false_after_logging_warning
+      Rails.logger.warn "Unknown form handler '#{handler}' (form ID: #{id})"
+      false
+    end
+
+    def handler_exists?
+      handlers.respond_to?( handler.to_sym )
```

Comment:
Could this allow an attacker to call one of the public methods `FormHandler` inherits from `Object` to be called?

If so, see the available public methods by running this in the rails console:

```rb
puts ShinyForm::FormHandler.new.public_methods
#puts Object.new.public_methods
```

Consider making this stricter by using `ShinyForm::FormHandler::FORM_HANDLERS ` as an allowlist.

Reply:
Good call, cheers. I didn't have the FORM_HANDLERS constant when I wrote this bit, I added that quite late in (and a bit reluctantly) because I couldn't see a neater way to generate the drop-down in the admin UI - unless I'd used the locale file as the data-source for there, which seems like a dodgy reversal of purpose. Now that I've got it though, definitely makes sense to use it here too.